pkg/lib/cockpit.js

```diff
@@ -2391,29 +2391,28 @@ function factory() {
 
     let the_user = null;
     cockpit.user = function () {
-        const dfd = cockpit.defer();
-        if (!the_user) {
-            const dbus = cockpit.dbus(null, { bus: "internal" });
-            dbus.call("/user", "org.freedesktop.DBus.Properties", "GetAll",
-                      ["cockpit.User"], { type: "s" })
-                .then(([user]) => {
-                    the_user = {
-                        id: user.Id.v,
-                        name: user.Name.v,
-                        full_name: user.Full.v,
-                        groups: user.Groups.v,
-                        home: user.Home.v,
-                        shell: user.Shell.v
-                    };
-                    dfd.resolve(the_user);
-                })
-                .catch(ex => dfd.reject(ex))
-                .finally(() => dbus.close());
-        } else {
-            dfd.resolve(the_user);
-        }
-
-        return dfd.promise;
+        return new Promise((resolve, reject) => {
+            if (!the_user) {
+                const dbus = cockpit.dbus(null, { bus: "internal" });
+                dbus.call("/user", "org.freedesktop.DBus.Properties", "GetAll",
+                          ["cockpit.User"], { type: "s" })
+                    .then(([user]) => {
+                        the_user = {
+                            id: user.Id.v,
+                            name: user.Name.v,
+                            full_name: user.Full.v,
+                            groups: user.Groups.v,
+                            home: user.Home.v,
+                            shell: user.Shell.v
+                        };
+                        resolve(the_user);
+                    })
+                    .catch(ex => reject(ex))
```

Comment:
... then you can also drop this boilerplate.

Reply:
Right, but how would the caching work then? Or do we drop that, it doesn't seem too useful..

Reply:
No, please keep it. I though just like this:
```js
if (the_user)
    return Promise.resolve(the_user);

return dbus.call(...).then([user] => {
    the_user = {...};
    return the_user;
});
```

Reply:
Tried that, but it just explodes.

```Uncaught TypeError: import_cockpit23.default.user() is undefined
    LastLogin lastLogin.jsx:77
```

As `cockpit.user()` no longer returns a promise:

```
    let the_user = null;
    cockpit.user = function () {
            if (!the_user) {
                const dbus = cockpit.dbus(null, { bus: "internal" });
                return dbus.call("/user", "org.freedesktop.DBus.Properties", "GetAll",
                          ["cockpit.User"], { type: "s" })
                    .then(([user]) => {
                        the_user = {
                            id: user.Id.v,
                            name: user.Name.v,
                            full_name: user.Full.v,
                            groups: user.Groups.v,
                            home: user.Home.v,
                            shell: user.Shell.v
                        };
                        return the_user;
                    })
                    .finally(() => dbus.close());
            } else {
                Promise.resolve(the_user);
            }
    };
```

```
cockpit.user()
undefined 
```

So I'm gonna leave this out till the DBus API uses a proper promise API

Reply:
Yes, you forgot the `return` in front of `Promise.resolve(the_user);` in the `else` branch :grin:

Reply:
Facepalm, re-pushed! Hope it goes green.